### PR TITLE
ECL: add partitioners

### PR DESCRIPTION
**Pull request by peterrum**
Related to issue #9153.

Ping @kronbichler

### CODE REVIEW

**Comment by kronbichler** [comment]
This looks alright, but I need some time to think about a proper name.

**Comment by peterrum** [comment]
@kronbichler `/rebuild` ?

File: include/deal.II/matrix_free/matrix_free.h [hunk 1]
```diff
@@ -725,6 +725,11 @@ class MatrixFree : public Subscriptor
      */
     values,
 
+    /**
+     * TODO
+     */
+    values_cell,
```
Comment (peterrum) [inline include/deal.II/matrix_free/matrix_free.h]: Maybe:
```suggestion
    values_all_faces
```

File: include/deal.II/matrix_free/matrix_free.templates.h [hunk 2]
```diff
@@ -1788,6 +1788,254 @@ MatrixFree<dim, Number, VectorizedArrayType>::initialize_indices(
...
+                ghost_indices.erase(std::unique(ghost_indices.begin(),
+                                                ghost_indices.end()),
+                                    ghost_indices.end());
+                IndexSet compressed_set(part.size());
+                compressed_set.add_indices(ghost_indices.begin(),
+                                           ghost_indices.end());
+                compressed_set.subtract_set(
+                  di.vector_partitioner->locally_owned_range());
+                const bool all_ghosts_equal = Utilities::MPI::min(
+                  (int)(compressed_set.n_elements() ==
```
Comment (masterleinad) [inline include/deal.II/matrix_free/matrix_free.templates.h]: We don't use C-style casts.

File: include/deal.II/matrix_free/matrix_free.h (hunk 1 above)
Comment (kronbichler) [reply include/deal.II/matrix_free/matrix_free.h]: Excellent choice, yes, please use `values_all_faces` and `gradients_all_faces`.

File: include/deal.II/matrix_free/matrix_free.templates.h [hunk 3]
```diff
@@ -1788,6 +1788,254 @@ MatrixFree<dim, Number, VectorizedArrayType>::initialize_indices(
                   }
               }
           }
+
+
+          // partitioner 3: values on faces
```
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.templates.h]: What comes next is quite a lot of code. Do you see a chance to combine it with the other code somehow, e.g. by a separate function or by a lambda? I have not looked into it in detail yet, but 90% of code seems very similar.

Comment (peterrum) [reply include/deal.II/matrix_free/matrix_free.templates.h]: I will try to introduce a lambda.

**Comment by masterleinad** [comment]
```
The following tests FAILED:
	1992 - matrix_free/interpolate_quadratic.debug (Failed)
	2015 - matrix_free/matrix_vector_faces_05.debug (Failed)
	2026 - matrix_free/interpolate_constant.debug (Failed)
	2071 - matrix_free/interpolate_linear.debug (Failed)
	2101 - matrix_free/interpolate_cubic.debug (Failed)
```

**Comment by peterrum** [comment]
@kronbichler As promised, I have introduced the lambdas. Do like it better now?

(I know I have to spell out the autos; I'll do it if you like it as it is!)

File: include/deal.II/matrix_free/matrix_free.templates.h [hunk 4]
```diff
@@ -1589,205 +1589,304 @@ MatrixFree<dim, Number, VectorizedArrayType>::initialize_indices(
...
+                  &vector_partitoner_values,
+                std::shared_ptr<const Utilities::MPI::Partitioner>
+                  &vector_partitioner_gradients,
+                const std::function<void(
+                  const std::function<void(const unsigned int,
+                                           const unsigned int,
+                                           const unsigned int,
+                                           const unsigned int,
+                                           const bool,
+                                           const bool)> &)> &loop) {
```
Comment (peterrum) [inline include/deal.II/matrix_free/matrix_free.templates.h]: @kronbichler  Beautiful ;) I hope we never switch to C++14

**Comment by kronbichler** [comment]
Can you have a look at the serial build error
```
In file included from /jenkins/workspace/dealii_PR-9156/source/matrix_free/matrix_free.cc:21:0:
/jenkins/workspace/dealii_PR-9156/include/deal.II/matrix_free/matrix_free.templates.h: In lambda function:
/jenkins/workspace/dealii_PR-9156/include/deal.II/matrix_free/matrix_free.templates.h:1685:46: error: default argument specified for lambda parameter [-Werror=pedantic]
                            const bool flag = false) {
                                              ^
/jenkins/workspace/dealii_PR-9156/include/deal.II/matrix_free/matrix_free.templates.h: In lambda function:
/jenkins/workspace/dealii_PR-9156/include/deal.II/matrix_free/matrix_free.templates.h:1796:54: error: default argument specified for lambda parameter [-Werror=pedantic]
                            const bool         flag = false) {
                                                      ^
```
Then this should be ready to be merged.

**Pull request merged**